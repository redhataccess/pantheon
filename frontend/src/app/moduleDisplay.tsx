import React, { Component } from 'react';
import { Level, LevelItem, Breadcrumb, BreadcrumbItem } from '@patternfly/react-core';
import {
    DataList, DataListItem, DataListItemRow, DataListItemCells,
    DataListCell, Card, Text, TextContent, TextVariants
} from '@patternfly/react-core';
import { Versions } from '@app/versions';

class ModuleDisplay extends Component<any, any, any> {

    constructor(props) {
        super(props)
        this.state = {
            copySuccess: '',
            draftPath: '',
            draftUpdateDate: '',
            modulePath: '',
            moduleTitle: "",
            moduleType: '',
            moduleUUID: '',
            productValue: "",
            releasePath: '',
            releaseUpdateDate: '',
            releaseVersion: '',
            results: {},
            versionUUID: "",
            versionValue: ""
        };
    }

    public componentDidMount() {
        this.fetchModuleDetails(this.props)
        this.getVersionUUID(this.props.location.pathname)
    }

    public render() {
        // console.log('Props: ', this.props);
        return (
            <React.Fragment>
                <div>
                    <div className="app-container">
                        <Breadcrumb>
                            <BreadcrumbItem ><a href="#/search">Modules</a></BreadcrumbItem>
                            <BreadcrumbItem to="#" isActive={true}>{this.state.moduleTitle}</BreadcrumbItem>
                        </Breadcrumb>
                    </div>
                    <br />
                    <div>
                        <Level gutter="md">
                            <LevelItem>
                                <TextContent>
                                    <Text component={TextVariants.h1}>{this.state.moduleTitle}</Text>
                                </TextContent>
                            </LevelItem>
                            <LevelItem />
                        </Level>
                    </div>
                    <div>
                        {this.state.releaseUpdateDate.trim() !== "" && this.state.releaseUpdateDate !== '-'
                            && this.state.moduleUUID !== ""
                            && <span><a href={'https://access.redhat.com/topics/en-us/' + this.state.moduleUUID} target="_blank">View on Customer Portal  <i className="fa pf-icon-arrow" /></a> </span>
                        }

                        <span>&emsp;&emsp;</span>

                        {this.state.releaseUpdateDate.trim() !== "" && this.state.releaseUpdateDate !== '-'
                            && this.state.moduleUUID !== ""
                            && <span><a id="permanentURL" onClick={this.copyToClipboard}>Copy permanent URL  <i className="fa pf-icon-folder-close" /></a></span>
                        }

                        <span>&emsp;{this.state.copySuccess !== '' && this.state.copySuccess}</span>
                    </div>
                    <br />
                    <div>
                        <DataList aria-label="single action data list">
                            <DataListItem aria-labelledby="simple-item1">
                                <DataListItemRow id="data-rows-header" >
                                    <DataListItemCells
                                        dataListCells={[
                                            <DataListCell width={2} key="product">
                                                <span className="sp-prop-nosort" id="span-source-type-product">Product</span>
                                            </DataListCell>,
                                            <DataListCell key="published">
                                                <span className="sp-prop-nosort" id="span-source-type-published">Published</span>
                                            </DataListCell>,
                                            <DataListCell key="updated">
                                                <span className="sp-prop-nosort" id="span-source-type-draft-uploaded">Draft Uploaded</span>
                                            </DataListCell>,
                                            <DataListCell key="module_type">
                                                <span className="sp-prop-nosort" id="span-source-name-module-type">Module Type</span>
                                            </DataListCell>
                                        ]}
                                    />
                                </DataListItemRow>
                                <DataListItemRow>
                                    <DataListItemCells
                                        dataListCells={[
                                            <DataListCell width={2} key="product">
                                                <span>{this.state.productValue + ' ' + this.state.versionValue}</span>
                                            </DataListCell>,
                                            <DataListCell key="published">
                                                <span>
                                                    {this.state.releaseUpdateDate.trim() !== ""
                                                        && this.state.releaseUpdateDate.length >= 15 ?
                                                        this.state.releaseUpdateDate : "-"} <br />
                                                    {this.state.releaseUpdateDate.trim() !== "" &&
                                                        <a href={this.state.releasePath} target="_blank">{this.state.releaseVersion}</a>}
                                                </span>
                                            </DataListCell>,
                                            <DataListCell key="updated">
                                                <span>
                                                    {this.state.draftUpdateDate.trim() !== ""
                                                        && this.state.draftUpdateDate.length >= 15 ?
                                                        this.state.draftUpdateDate : "-"}
                                                </span>
                                            </DataListCell>,
                                            <DataListCell key="module_type">
                                                <span>{this.state.moduleType !== undefined ? this.state.moduleType : "-"}</span>
                                            </DataListCell>,
                                        ]}
                                    />
                                </DataListItemRow>
                            </DataListItem>
                        </DataList>
                    </div>
                    <div>
                        <Card>
                            <Versions
                                modulePath={this.state.modulePath}
                                versionModulePath={this.state.moduleTitle}
                                updateDate={this.updateDate}
                                onGetProduct={this.getProduct}
                                onGetVersion={this.getVersion}
                            />
                        </Card>
                    </div>
                </div>
            </React.Fragment>
        );
    }

    private updateDate = (draftDate, releaseDate, releaseVersion, moduleUUID) => {
        this.setState({
            draftUpdateDate: draftDate,
            moduleUUID,
            releaseUpdateDate: releaseDate,
            releaseVersion,
        });
    }

    private fetchModuleDetails = (data) => {
        this.setState({
            modulePath: data.location.pathname,
            releasePath: "/content" + data.location.pathname + ".preview"
        })

        fetch(data.location.pathname + '.4.json')
            .then(response => response.json())
            .then(responseJSON => {
                // console.log('fetch results:', responseJSON)
                this.setState({
                    moduleTitle: responseJSON.en_US["1"].metadata["jcr:title"],
                    moduleType: responseJSON.en_US["1"].metadata["pant:moduleType"],

                })
            })
    }

    private getProduct = (product) => {
        // console.log("[getProduct] ", product)
        this.setState({ productValue: product })
    }

    private getVersion = (version) => {
        // console.log("[getVersion] ", version)
        this.setState({ versionValue: version })
    }

    private getVersionUUID = (path) => {
        // console.log("[getVersionUUID] path", path)
        path = "/content" + path + "/en_US/1/metadata.json"
        fetch(path)
            .then(response => response.json())
            .then((responseJSON) => {
                // console.log("[responseJSON]", responseJSON)
                if (responseJSON.productVersion !== undefined) {
                    // console.log("[allproducts inside fetch call] ", this.state.allProducts)

                    // const versionUUID = responseJSON["productVersion"]
                    this.setState({ versionUUID: responseJSON.productVersion }, () => {
                        this.getProductInitialLoad()
                    })
                }
            })

    }

    private getProductInitialLoad = () => {
        const path = '/content/products.3.json'
        let key
        fetch(path)
            .then(response => response.json())
            .then(responseJSON => {
                // console.log('fetch results:',responseJSON)
                // tslint:disable-next-line: prefer-for-of
                for (let i = 0; i < Object.keys(responseJSON).length; i++) {
                    key = Object.keys(responseJSON)[i];
                    const versionKey = "versions"
                    if ((key !== 'jcr:primaryType')) {
                        if (responseJSON[key].name !== undefined) {
                            const pName = responseJSON[key].name
                            const versionObj = responseJSON[key][versionKey]
                            if (versionObj) {
                                // console.log("[getProductFromUUID] versionObj ", versionObj)
                                let vKey;
                                const versions = new Array();
                                // tslint:disable-next-line: no-shadowed-variable
                                for (const item in Object.keys(versionObj)) {
                                    if (Object.keys(versionObj)[item] !== undefined) {
                                        vKey = Object.keys(versionObj)[item]
                                        if (vKey !== 'jcr:primaryType') {
                                            if (versionObj[vKey].name) {
                                                if (versionObj[vKey]["jcr:uuid"] === this.state.versionUUID) {
                                                    // process productValue and versionValue on initial load
                                                    this.setState({ productValue: pName, versionValue: versionObj[vKey].name }, () => {
                                                        // console.log("[getProductFromUUID] item/productvalue", pName)
                                                        // console.log("[getProductFromUUID] versionValue", versionObj[vKey][nameKey])
                                                    })
                                                    break;
                                                }

                                            }
                                        }
                                    }
                                }

                            }

                        }
                    }
                }

            })
    }

    private copyToClipboard = () => {
        const textField = document.createElement('textarea')
        if (window.location.href !== undefined) {
            const targetHref = window.location.href
            if (window.location.pathname !== undefined) {
                textField.innerText = targetHref.split(window.location.pathname)[0] + this.state.releasePath
                document.body.appendChild(textField)
                textField.select()
                document.execCommand('copy')
                textField.remove()
                this.setState({ copySuccess: 'Copied!' });
            }
        }
    };
}

export { ModuleDisplay }