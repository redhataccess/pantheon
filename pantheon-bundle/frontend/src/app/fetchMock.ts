if (!self.fetch) {
    self.fetch = (input: RequestInfo, init?: RequestInit | undefined) => {
        const p = new Promise<Response>((resolve, reject) => {
            // just don't do anything. this exists so that our tests don't try and actually resolve fetches,
            // (which is exactly what isomorphic-fetch tried to do while we were importing it...)
            // and calling nothing here allows us to skip all of our fetch.then() and fetch.catch() code.
        })
        return p
    }
}