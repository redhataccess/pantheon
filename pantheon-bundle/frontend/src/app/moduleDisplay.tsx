import React, { Component } from 'react'
import { CopyIcon } from '@patternfly/react-icons';
import { Level, LevelItem, Breadcrumb, BreadcrumbItem, Button } from '@patternfly/react-core'
import {
    DataList, DataListItem, DataListItemRow, DataListItemCells,
    DataListCell, Card, Text, TextContent, TextVariants
} from '@patternfly/react-core'
import { Versions } from '@app/versions'
import { Fields } from '@app/Constants'
import { continueStatement } from '@babel/types';

class ModuleDisplay extends Component<any, any, any> {

    constructor(props) {
        super(props)
        this.state = {
            copySuccess: '',
            draftPath: '',
            draftUpdateDate: '',
            modulePath: '',
            moduleTitle: "",
            moduleType: '',
            variantUUID: '',
            portalHost: '',
            productValue: "",
            releasePath: '',
            releaseUpdateDate: '',
            releaseVersion: '',
            results: {},
            variant: 'DEFAULT',
            versionValue: ""
        }
    }

    public componentDidMount() {
        // this.getVariantParam()
        this.fetchModuleDetails(this.props)
        this.getVersionUUID(this.props.location.pathname)
        this.getPortalUrl()
    }

    public render() {
        // console.log('Props: ', this.props)
        return (
            <React.Fragment>
                <div>
                    <div className="app-container">
                        <Breadcrumb>
                            <BreadcrumbItem ><a href="#/search">Modules</a></BreadcrumbItem>
                            <BreadcrumbItem to="#" isActive={true}>{this.state.moduleTitle}</BreadcrumbItem>
                        </Breadcrumb>
                    </div>
                    <br />
                    <div>
                        <Level gutter="md">
                            <LevelItem>
                                <TextContent>
                                    <Text component={TextVariants.h1}>{this.state.moduleTitle}</Text>
                                </TextContent>
                            </LevelItem>
                            <LevelItem />
                        </Level>
                    </div>
                    <div>
                        {this.state.releaseUpdateDate.trim() !== "" && this.state.releaseUpdateDate !== '-'
                            && this.state.variantUUID !== ""
                            && this.state.portalHost !== ""
                            && <span><a href={this.state.portalHost + '/topics/en-us/' + this.state.variantUUID} target="_blank">View on Customer Portal  <i className="fa pf-icon-arrow" /></a> </span>
                        }

                        <span>&emsp;&emsp;</span>

                        {this.state.releaseUpdateDate.trim() !== "" && this.state.releaseUpdateDate !== '-'
                            && this.state.variantUUID !== ""
                            && this.state.portalHost !== ""
                            && <span><a id="permanentURL" onClick={this.copyToClipboard} onMouseLeave={this.mouseLeave}>Copy permanent URL  <CopyIcon /></a></span>
                        }

                        <span>&emsp;{this.state.copySuccess !== '' && this.state.copySuccess}</span>
                    </div>
                    <br />
                    <Level gutter="md">
                        <LevelItem>{}</LevelItem>
                        <LevelItem>{}</LevelItem>
                        <LevelItem>
                            <Button variant='primary' onClick={() => this.generateDraftHtml(this.props.location.pathname)}>Generate Draft Html</Button>{'  '}
                        </LevelItem>
                    </Level>
                    <br />
                    <div>
                        <DataList aria-label="single action data list">
                            <DataListItem aria-labelledby="simple-item1">
                                <DataListItemRow id="data-rows-header" >
                                    <DataListItemCells
                                        dataListCells={[
                                            <DataListCell width={2} key="product">
                                                <span className="sp-prop-nosort" id="span-source-type-product">Product</span>
                                            </DataListCell>,
                                            <DataListCell key="published">
                                                <span className="sp-prop-nosort" id="span-source-type-published">Published</span>
                                            </DataListCell>,
                                            <DataListCell key="updated">
                                                <span className="sp-prop-nosort" id="span-source-type-draft-uploaded">Draft Uploaded</span>
                                            </DataListCell>,
                                            <DataListCell key="module_type">
                                                <span className="sp-prop-nosort" id="span-source-name-module-type">Module Type</span>
                                            </DataListCell>
                                        ]}
                                    />
                                </DataListItemRow>
                                <DataListItemRow>
                                    <DataListItemCells
                                        dataListCells={[
                                            <DataListCell width={2} key="product">
                                                <span>{this.state.productValue + ' ' + this.state.versionValue}</span>
                                            </DataListCell>,
                                            <DataListCell key="published">
                                                <span>
                                                    {this.state.releaseUpdateDate.trim() !== ""
                                                        && this.state.releaseUpdateDate.length >= 15 ?
                                                        this.state.releaseUpdateDate : "-"} <br />
                                                    {this.state.releaseUpdateDate.trim() !== "" &&
                                                        <a href={this.state.releasePath} target="_blank">{this.state.releaseVersion}</a>}
                                                </span>
                                            </DataListCell>,
                                            <DataListCell key="updated">
                                                <span>
                                                    {this.state.draftUpdateDate.trim() !== ""
                                                        && this.state.draftUpdateDate.length >= 15 ?
                                                        this.state.draftUpdateDate : "-"}
                                                </span>
                                            </DataListCell>,
                                            <DataListCell key="module_type">
                                                <span>{this.state.moduleType !== undefined ? this.state.moduleType : "-"}</span>
                                            </DataListCell>,
                                        ]}
                                    />
                                </DataListItemRow>
                            </DataListItem>
                        </DataList>
                    </div>
                    <div>
                        <Card>
                            <Versions
                                modulePath={this.state.modulePath}
                                productInfo={this.state.productValue}
                                versionModulePath={this.state.moduleTitle}
                                variant={this.state.variant}
                                updateDate={this.updateDate}
                                onGetProduct={this.getProduct}
                                onGetVersion={this.getVersion}
                            />
                        </Card>
                    </div>
                </div>
            </React.Fragment>
        )
    }
    private generateDraftHtml = (pathname: any) => {
        const docPath = '/content' + pathname + '.preview?draft=true&variant=' + this.state.variant

        // console.log('Preview path: ', docPath)
        return window.open(docPath)
    }

    private updateDate = (draftDate, releaseDate, releaseVersion, variantUUID) => {
        this.setState({
            draftUpdateDate: draftDate,
            variantUUID,
            releaseUpdateDate: releaseDate,
            releaseVersion,
        })
    }

    private fetchModuleDetails = async(data) => {
        await this.getVariantParam()
        this.setState({
            modulePath: data.location.pathname,
            releasePath: "/content" + data.location.pathname + ".preview?variant=" + this.state.variant
        })

        fetch(data.location.pathname + '/en_US.harray.4.json')
            .then(response => response.json())
            .then(responseJSON => {
                // console.log('fetch results:', responseJSON)
                // TODO: refactor for loops
                for (const sourceVariant of responseJSON.__children__) {
                    if (!sourceVariant.__children__) {
                        continue
                    }
                    for (const myChild of sourceVariant.__children__) {

                        if (!myChild.__children__) {
                            continue
                        }
                        if (myChild.__name__ === 'draft') {

                            this.setState({ draftUpdateDate: myChild["jcr:created"] })
                        }
                        for (const myGrandchild of myChild.__children__) {
                            if (!myGrandchild.__children__) {
                                continue
                            }

                            for (const offspring of myGrandchild.__children__) {
                                if (offspring.__name__ === 'metadata') {

                                    if (offspring[Fields.JCR_TITLE] !== undefined) {

                                        this.setState({
                                            moduleTitle: offspring[Fields.JCR_TITLE],
                                        })
                                    }
                                    if (offspring[Fields.PANT_MODULE_TYPE] !== undefined) {

                                        this.setState({
                                            moduleType: offspring[Fields.PANT_MODULE_TYPE],

                                        })
                                    }
                                }
                            }

                        }
                    }
                }

            })
    }

    private getProduct = (product) => {
        this.setState({ productValue: product })
    }

    private getVersion = (version) => {
        this.setState({ versionValue: version })
    }

    private getVersionUUID = (path) => {
        // path = "/content" + path + "/en_US/1/metadata.json"
        path = "/content" + path + "/en_US.harray.4.json"
        fetch(path)
            .then(response => response.json())
            .then((responseJSON) => {
                for (const locale of responseJSON.__children__) {
                    if (!locale.__children__) {
                        continue
                    }
                    for (const localeChild of locale.__children__) {

                        if (!localeChild.__children__) {
                            continue
                        }
                        for (const variant of localeChild.__children__) {
                            if (!variant.__children__) {
                                continue
                            }

                            for (const offspring of variant.__children__) {
                                if (offspring.__name__ === 'metadata') {

                                    if (offspring[Fields.PANT_PRODUCT_VERSION_REF] !== undefined) {

                                        this.getProductInitialLoad(offspring[Fields.PANT_PRODUCT_VERSION_REF])
                                    }
                                }
                            }

                        }
                    }
                }
            })
    }

    private getProductInitialLoad = (uuid) => {
        const path = '/content/products.harray.3.json'
        fetch(path)
            .then(response => response.json())
            .then(responseJSON => {
                for (const product of responseJSON.__children__) {
                    if (!product.__children__) {
                        continue
                    }
                    for (const productChild of product.__children__) {
                        if (productChild.__name__ !== 'versions') {
                            continue
                        }
                        for (const productVersion of productChild.__children__) {
                            if (productVersion[Fields.JCR_UUID] === uuid) {
                                this.setState({ productValue: product.name, versionValue: productVersion.name })
                                break
                            }
                        }
                    }
                }
            })
    }

    private copyToClipboard = () => {
        const textField = document.createElement('textarea')
        if (this.state.variantUUID.trim() !== '') {
            textField.value = this.state.portalHost + '/topics/en-us/' + this.state.variantUUID
            document.body.appendChild(textField)
            textField.select()
            document.execCommand('copy')
            textField.remove()
            this.setState({ copySuccess: 'Copied!' })
        }
    }

    private mouseLeave = () => {
        this.setState({ copySuccess: '' })
    }

    private getPortalUrl = () => {
        fetch('/conf/pantheon/pant:portalUrl')
            .then(resp => {
                if (resp.ok) {
                    resp.text().then(text => {
                        this.setState({ portalHost: text })
                        // console.log("set portalHost: " + this.state.portalHost)
                    })
                }
            })
    }

    private async getVariantParam() {
        const query = new URLSearchParams(this.props.location.search);
        const variantParam = query.get('variant')
        // console.log("[moduleDisplay] variantParam => ", variantParam)
        if (variantParam !== 'undefined') {
            this.setState({ variant: variantParam })
        }
    }

}

export { ModuleDisplay }
