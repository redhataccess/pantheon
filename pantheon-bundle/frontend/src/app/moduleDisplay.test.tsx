import React from 'react'
import { ModuleDisplay } from '@app/moduleDisplay'
import '@app/fetchMock'

import { mount, shallow } from 'enzyme'
import { Button, Card, DataList, DataListItem, DataListItemCells, DataListItemRow, DataListCell, TextContent, Level, LevelItem, Breadcrumb, BreadcrumbItem } from '@patternfly/react-core'
import renderer from 'react-test-renderer'
import sinon from 'sinon'
import { Versions } from '@app/versions'
const anymatch = require('anymatch')

const props = {
    location: { pathname: "module/test" }
}

describe('ModuleDisplay tests', () => {
    test('should render ModuleDisplay component', () => {
        const view = shallow(<ModuleDisplay {...props} />)
        expect(view).toMatchSnapshot()
    })

    it('should render a Breadcrumb', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const breadcrumb = wrapper.find(Breadcrumb)
        expect(breadcrumb.exists()).toBe(true)
    })

    it('should render a BreadcrumbItem', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const breadcrumbItem = wrapper.find(BreadcrumbItem)
        expect(breadcrumbItem.exists()).toBe(true)
    })

    it('should render a Button', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const button = wrapper.find(Button)
        expect(button.exists()).toBe(true)
    })

    it('should render a Card', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const card = wrapper.find(Card)
        expect(card.exists()).toBe(true)
    })

    it('should render a Data List', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const dataList = wrapper.find(DataList)
        expect(dataList.exists()).toBe(true)
    })

    it('should render a DataListItem', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const dataListItem = wrapper.find(DataListItem)
        expect(dataListItem.exists()).toBe(true)
    })

    it('should render a DataListItemCells Element', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const dataListItemCells = wrapper.find(DataListItemCells)
        expect(dataListItemCells.exists()).toBe(true)
    })

    it('should render a DataListItemRow element', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const dataListItemRow = wrapper.find(DataListItemRow)
        expect(dataListItemRow.exists()).toBe(true)
    })

    it('should render a DataListCell', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const dataListCell = wrapper.find(DataListCell)
        expect(dataListCell.exists()).toBe(true)
        // console.log("[DataListCell] length", dataListCell.length)
        expect(dataListCell.at(0).contains("Product")).toBe(true)
    })

    it('should render a TextContent Element', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const textContent = wrapper.find(TextContent)
        expect(textContent.exists()).toBe(true)
    })

    it('should render a Level Element', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const level = wrapper.find(Level)
        expect(level.exists()).toBe(true)
    })

    it('should render a LevelItem', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const levelItem = wrapper.find(LevelItem)
        expect(levelItem.exists()).toBe(true)
    })

    it('should render a Versions Element', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const versions = wrapper.find(Versions)
        expect(versions.exists()).toBe(true)
    })

    it('should render a h1 component', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const h1 = wrapper.find('[component="h1"]')
        expect(h1.exists()).toBe(true)
    })

    it('test fetchModuleDetails function', () => {
        const wrapper = renderer.create(<ModuleDisplay {...props} />)
        const inst = wrapper.getInstance()
        expect(inst.fetchModuleDetails(props)).toMatchSnapshot()
    })

    it('test getProduct function', () => {
        const wrapper = renderer.create(<ModuleDisplay {...props} />)
        const inst = wrapper.getInstance()
        expect(inst.getProduct()).toMatchSnapshot()
    })

    it('test getVersion function', () => {
        const wrapper = renderer.create(<ModuleDisplay {...props} />)
        const inst = wrapper.getInstance()
        expect(inst.getVersion()).toMatchSnapshot()
    })

    it('test getVersionUUID function', () => {
        const wrapper = renderer.create(<ModuleDisplay {...props} />)
        const inst = wrapper.getInstance()
        expect(inst.getVersionUUID("/modules/test")).toMatchSnapshot()
    })

    it('test getProductInitialLoad function', () => {
        const wrapper = renderer.create(<ModuleDisplay {...props} />)
        const inst = wrapper.getInstance()
        expect(inst.getProductInitialLoad()).toMatchSnapshot()
    })

    it('test componentDidMount function', () => {
        const wrapper = renderer.create(<ModuleDisplay {...props} />)
        const inst = wrapper.getInstance()
        expect(inst.componentDidMount()).toMatchSnapshot()
    })

    it('test getPortalUrl function', () => {
        const wrapper = renderer.create(<ModuleDisplay {...props} />)
        const inst = wrapper.getInstance()
        expect(inst.getPortalUrl()).toMatchSnapshot()
    })

    it('has a props', () => {
        const moduleDisplay = mount(<ModuleDisplay {...props} />).matchesElement
        expect(moduleDisplay.length === 1)
    })

    // Value testing with Enzyme.
    it('renders Product heading', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const sourceTypeText = wrapper.find('#span-source-type-product').first().text()

        // ensure it matches what is expected
        expect(sourceTypeText).toEqual("Product")
    })

    it('renders Published heading', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const sourceTypeText = wrapper.find('#span-source-type-published').first().text()

        // ensure it matches what is expected
        expect(sourceTypeText).toEqual("Published")
    })

    it('renders Draft Uploaded heading', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const sourceTypeText = wrapper.find('#span-source-type-draft-uploaded').first().text()

        // ensure it matches what is expected
        expect(sourceTypeText).toEqual("Draft Uploaded")
    })

    it('renders Module Type heading', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        const sourceTypeText = wrapper.find('#span-source-name-module-type').first().text()

        // ensure it matches what is expected
        expect(sourceTypeText).toEqual("Module Type")
    })

    it('renders View on Customer Portal hotlink', () => {
        const wrapper = mount(<ModuleDisplay {...props} />)
        wrapper.setState({ 'login': true })
        wrapper.setState({ 'releaseUpdateDate': "Fri Oct 18 2019 17:35:50 GMT-0400" })
        wrapper.setState({ 'moduleUUID': "123" })
        wrapper.setState({ 'portalHost': "https://example.com" })
        const sourceTypeText = wrapper.find('a').at(2).text()

        // ensure it matches what is expected
        expect(sourceTypeText).toContain("View on Customer Portal")
    })

    it('should check if draftUpdateDate exists', () => {
        const wrapper = shallow(<ModuleDisplay {...props} />)
        wrapper.setState({ 'login': true })
        wrapper.setState({ "draftUpdateDate": "abcd" })
        expect(wrapper.state('draftUpdateDate')).toBeDefined()
    })

    it('should check if releaseUpdateDate exists', () => {
        const wrapper = shallow(<ModuleDisplay {...props} />)
        wrapper.setState({ 'login': true })
        wrapper.setState({ "releaseUpdateDate": "abcd" })
        expect(wrapper.state('releaseUpdateDate')).toBeDefined()
    })

    it('should check if moduleType exists', () => {
        const wrapper = shallow(<ModuleDisplay {...props} />)
        wrapper.setState({ 'moduleType': '' })
        wrapper.setState({ "moduleType": "module" })
        expect(wrapper.state('moduleType')).toBeDefined()
    })

    it('should have a moduleTitle', () => {
        const wrapper = shallow(<ModuleDisplay {...props} />)
        wrapper.setState({ 'login': true })
        wrapper.setState({ "moduleTitle": "test title" })

        const sourceText = wrapper.find('[component="h1"]').first().html()
        // ensure it matches what is expected
        expect(wrapper.state('moduleTitle')).toBeDefined()
        expect(sourceText).toContain("test title")
    })

    it('should have a versionUUID', () => {
        const wrapper = shallow(<ModuleDisplay {...props} />)
        wrapper.setState({ 'login': true })
        wrapper.setState({ "versionUUID": "122234-1234-1234T" })
        expect(wrapper.state('versionUUID')).toBeDefined()
    })

    it('should have a productValue', () => {
        const wrapper = shallow(<ModuleDisplay {...props} />)
        wrapper.setState({ 'login': true })
        const len = wrapper.setState({ "productValue": "Red Hat Enterprise Linux" })
        expect(wrapper.state('productValue')).toBeDefined()
    })

    it('should have a versionValue', () => {
        const wrapper = shallow(<ModuleDisplay {...props} />)
        wrapper.setState({ 'login': true })
        const len = wrapper.setState({ "versionValue": "8.x" })
        expect(wrapper.state('versionValue')).toBeDefined()
    })

    it('renders Copy permanent URL', () => {
        const wrapper = shallow(<ModuleDisplay {...props} />)
        wrapper.setState({ 'login': true })
        wrapper.setState({ 'moduleUUID': "somepath" })
        const permanentURL = wrapper.find('a#permanentURL').first()
        expect(permanentURL.exists).toBeTruthy()
    })

    it('renders copySuccess Message', () => {
        const wrapper = shallow(<ModuleDisplay {...props} />)
        wrapper.setState({ 'login': true })
        wrapper.setState({ 'releasePath': "somepath" })
        wrapper.setState({ 'moduleUUID': "1234" })
        wrapper.setState({ "copySuccess": "Copied!" })
        expect(wrapper.state('copySuccess')).toContain("Copied!")
    })

    test('copyToClipboard click event', () => {
        const wrapper = shallow(<ModuleDisplay {...props} />)
        const instance = wrapper.instance()
        const spy = sinon.spy(instance, 'copyToClipboard')

        wrapper.setState({ moduleUUID: '1234', releasePath: 'yarn' })
        expect(wrapper.find('#permanentURL').exists())
    })

    it('test mouseLeave function', () => {
        const wrapper = renderer.create(<ModuleDisplay {...props} />)
        const inst = wrapper.getInstance()
        expect(inst.mouseLeave()).toMatchSnapshot()
    })

    it('test fetch api call for portalUrl', async () => {
        window.fetch = jest.fn().mockImplementation(async () => {
            return new Promise((resolve, reject) => {
                resolve({
                    ok: true,
                    status: 200,
                    json: () => new Promise((resolve, reject) => {
                        resolve({
                            "portalHost": "https://example.com",
                        })
                    })
                })
            })
            const wrapper = await shallow(<ModuleDisplay {...props} />)
            await wrapper.update()
            expect(wrapper.state('portalHost')).toBe("https://example.com")
        })
    })
})
