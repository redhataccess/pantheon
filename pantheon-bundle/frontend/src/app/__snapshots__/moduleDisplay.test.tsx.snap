// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleDisplay tests should render ModuleDisplay component 1`] = `
<Fragment>
  <div>
    <div
      className="app-container"
    >
      <Component>
        <BreadcrumbItem>
          <a
            href="#/search"
          >
            Modules
          </a>
        </BreadcrumbItem>
        <BreadcrumbItem
          isActive={true}
          to="#"
        />
      </Component>
    </div>
    <br />
    <div>
      <Level
        gutter="md"
      >
        <LevelItem>
          <TextContent>
            <Text
              component="h1"
            />
          </TextContent>
        </LevelItem>
        <LevelItem />
      </Level>
    </div>
    <div>
      <span>
          
      </span>
      <span>
         
      </span>
    </div>
    <br />
    <Level
      gutter="md"
    >
      <LevelItem />
      <LevelItem />
      <LevelItem>
        <Component
          onClick={[Function]}
          variant="primary"
        >
          Generate Draft Html!!
        </Component>
          
      </LevelItem>
    </Level>
    <br />
    <div>
      <DataList
        aria-label="single action data list"
      >
        <DataListItem
          aria-labelledby="simple-item1"
        >
          <DataListItemRow
            id="data-rows-header"
          >
            <DataListItemCells
              dataListCells={
                Array [
                  <DataListCell
                    width={2}
                  >
                    <span
                      className="sp-prop-nosort"
                      id="span-source-type-product"
                    >
                      Product
                    </span>
                  </DataListCell>,
                  <DataListCell>
                    <span
                      className="sp-prop-nosort"
                      id="span-source-type-published"
                    >
                      Published
                    </span>
                  </DataListCell>,
                  <DataListCell>
                    <span
                      className="sp-prop-nosort"
                      id="span-source-type-draft-uploaded"
                    >
                      Draft Uploaded
                    </span>
                  </DataListCell>,
                  <DataListCell>
                    <span
                      className="sp-prop-nosort"
                      id="span-source-name-module-type"
                    >
                      Module Type
                    </span>
                  </DataListCell>,
                ]
              }
            />
          </DataListItemRow>
          <DataListItemRow>
            <DataListItemCells
              dataListCells={
                Array [
                  <DataListCell
                    width={2}
                  >
                    <span>
                       
                    </span>
                  </DataListCell>,
                  <DataListCell>
                    <span>
                      -
                       
                      <br />
                    </span>
                  </DataListCell>,
                  <DataListCell>
                    <span>
                      -
                    </span>
                  </DataListCell>,
                  <DataListCell>
                    <span>
                      
                    </span>
                  </DataListCell>,
                ]
              }
            />
          </DataListItemRow>
        </DataListItem>
      </DataList>
    </div>
    <div>
      <Card>
        <Versions
          modulePath=""
          onGetProduct={[Function]}
          onGetVersion={[Function]}
          productInfo=""
          updateDate={[Function]}
          variant={null}
          versionModulePath=""
        />
      </Card>
    </div>
  </div>
</Fragment>
`;

exports[`ModuleDisplay tests test componentDidMount function 1`] = `undefined`;

exports[`ModuleDisplay tests test fetchModuleDetails function 1`] = `Promise {}`;

exports[`ModuleDisplay tests test getPortalUrl function 1`] = `undefined`;

exports[`ModuleDisplay tests test getProduct function 1`] = `undefined`;

exports[`ModuleDisplay tests test getProductInitialLoad function 1`] = `undefined`;

exports[`ModuleDisplay tests test getVersion function 1`] = `undefined`;

exports[`ModuleDisplay tests test getVersionUUID function 1`] = `undefined`;

exports[`ModuleDisplay tests test mouseLeave function 1`] = `undefined`;
